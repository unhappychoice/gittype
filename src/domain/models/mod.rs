pub mod challenge;
pub mod chunk;
pub mod color_mode;
pub mod color_scheme;
pub mod config;
pub mod countdown;
pub mod extraction_options;
pub mod git_repository;
pub mod git_repository_ref;
pub mod language;
pub mod languages;
pub mod rank;
pub mod session;
pub mod stage;
pub mod storage;
pub mod theme;
pub mod total;
pub mod version;

// Re-export main types for easy access
pub use challenge::Challenge;
pub use chunk::{ChunkType, CodeChunk};
pub use countdown::Countdown;
pub use extraction_options::ExtractionOptions;
pub use git_repository::GitRepository;
pub use git_repository_ref::GitRepositoryRef;
pub use language::{Language, Languages};
pub use rank::{Rank, RankTier};
pub use session::{Session, SessionResult};
pub use stage::{Stage, StageResult};
pub use total::{Total, TotalResult};
