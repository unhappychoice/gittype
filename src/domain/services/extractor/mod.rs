pub mod challenge_converter;
pub mod code_chunk_extractor;
pub mod core;
pub mod git_repository_extractor;
pub mod language_registry;
pub mod parsers;
pub mod repository_extractor;

pub use challenge_converter::ChallengeConverter;
pub use code_chunk_extractor::CodeChunkExtractor;
pub use git_repository_extractor::GitRepositoryExtractor;
pub use language_registry::LanguageRegistry;
pub use repository_extractor::RepositoryExtractor;
