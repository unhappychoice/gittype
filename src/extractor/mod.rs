pub mod challenge_converter;
pub mod code_chunk_extractor;
pub mod core;
pub mod git_repository_extractor;
pub mod models;
pub mod parsers;
pub mod repository_extractor;

pub use crate::game::screens::loading_screen::{NoOpProgressReporter, ProgressReporter};
pub use crate::models::{ChunkType, CodeChunk, GitRepository};
pub use challenge_converter::ChallengeConverter;
pub use code_chunk_extractor::CodeChunkExtractor;
pub use git_repository_extractor::GitRepositoryExtractor;
pub use models::{ExtractionOptions, Language};
pub use repository_extractor::RepositoryExtractor;
