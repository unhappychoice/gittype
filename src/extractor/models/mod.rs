pub mod chunk;
pub mod language;
pub mod options;

pub use chunk::{ChunkType, CodeChunk};
pub use language::Language;
pub use options::ExtractionOptions;
