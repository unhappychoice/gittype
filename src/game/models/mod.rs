pub mod loading_steps;

pub use loading_steps::*;
