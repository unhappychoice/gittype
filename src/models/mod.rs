pub mod chunk;
pub mod challenge;
pub mod stage;
pub mod session;
pub mod total;
pub mod rank;
pub mod git_repository;

// Re-export main types for easy access
pub use chunk::{ChunkType, CodeChunk};
pub use challenge::Challenge;
pub use stage::{Stage, StageResult};
pub use session::{Session, SessionResult};
pub use total::{Total, TotalResult};
pub use rank::{Rank, RankingTitle};
pub use git_repository::GitRepository;