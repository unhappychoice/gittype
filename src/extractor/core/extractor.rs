use crate::extractor::parsers::{get_parser_registry, parse_with_thread_local};
use crate::models::CodeChunk;
use crate::{GitTypeError, Result};
use std::fs;
use std::path::Path;
use streaming_iterator::StreamingIterator;
use tree_sitter::{Node, QueryCursor, Tree};

pub struct CommonExtractor;

impl CommonExtractor {
    pub fn extract_chunks_from_tree(
        tree: &Tree,
        source_code: &str,
        file_path: &Path,
        language: &str,
    ) -> Result<Vec<CodeChunk>> {
        let mut chunks = Vec::new();
        let registry = get_parser_registry();

        let file_comment_ranges = Self::extract_comment_ranges(tree, source_code, language)?;
        let query = registry.create_query(language)?;

        let mut cursor = QueryCursor::new();
        let mut matches = cursor.matches(&query, tree.root_node(), source_code.as_bytes());

        while let Some(match_) = matches.next() {
            for capture in match_.captures {
                let node = capture.node;
                let capture_name = &query.capture_names()[capture.index as usize];

                if let Some(chunk) = Self::node_to_chunk(
                    node,
                    source_code,
                    file_path,
                    language,
                    capture_name,
                    &file_comment_ranges,
                ) {
                    chunks.push(chunk);
                }
            }
        }

        Ok(chunks)
    }

    pub fn extract_from_file(file_path: &Path, language: &str) -> Result<Vec<CodeChunk>> {
        let content = fs::read_to_string(file_path)?;
        // Reuse per-thread parser instance for the language
        let tree = parse_with_thread_local(language, &content).ok_or_else(|| {
            GitTypeError::ExtractionFailed(format!("Failed to parse file: {:?}", file_path))
        })?;

        Self::extract_chunks_from_tree(&tree, &content, file_path, language)
    }

    pub fn extract_comment_ranges(
        tree: &Tree,
        source_code: &str,
        language: &str,
    ) -> Result<Vec<(usize, usize)>> {
        let registry = get_parser_registry();
        let comment_query = registry.create_comment_query(language)?;
        let mut comment_ranges = Vec::new();

        let mut cursor = QueryCursor::new();
        let mut matches = cursor.matches(&comment_query, tree.root_node(), source_code.as_bytes());

        while let Some(m) = matches.next() {
            for capture in m.captures {
                let node = capture.node;
                let start = node.start_byte();
                let end = node.end_byte();

                if Self::is_valid_comment_node(node, language) {
                    comment_ranges.push((start, end));
                }
            }
        }

        comment_ranges.sort_by_key(|&(start, _)| start);
        Ok(comment_ranges)
    }

    fn is_valid_comment_node(node: Node, language: &str) -> bool {
        let node_kind = node.kind();
        match language {
            "rust" => node_kind == "line_comment" || node_kind == "block_comment",
            "typescript" => node_kind == "comment",
            "javascript" => node_kind == "comment",
            "python" => node_kind == "comment",
            "ruby" => node_kind == "comment",
            "go" => node_kind == "comment",
            "swift" => node_kind == "comment" || node_kind == "multiline_comment",
            "kotlin" => node_kind == "line_comment" || node_kind == "multiline_comment",
            "java" => node_kind == "line_comment" || node_kind == "block_comment",
            "php" => node_kind == "comment" || node_kind == "shell_comment_line",
            "csharp" => node_kind == "comment",
            "c" => node_kind == "comment",
            "cpp" => node_kind == "comment",
            "haskell" => node_kind == "comment",
            "dart" => node_kind == "comment" || node_kind == "documentation_comment",
            _ => false,
        }
    }

    fn node_to_chunk(
        node: Node,
        source_code: &str,
        file_path: &Path,
        language: &str,
        capture_name: &str,
        file_comment_ranges: &[(usize, usize)],
    ) -> Option<CodeChunk> {
        let start_byte = node.start_byte();
        let end_byte = node.end_byte();
        let content = &source_code[start_byte..end_byte];

        let start_line = node.start_position().row + 1;
        let end_line = node.end_position().row + 1;
        let original_indentation = node.start_position().column;

        // Extract actual indentation characters from source
        let original_indent_chars = if original_indentation > 0 {
            Self::extract_line_indent_chars(
                source_code,
                node.start_position().row,
                original_indentation,
            )
        } else {
            String::new()
        };

        let registry = get_parser_registry();
        let extractor = registry.get_extractor(language).ok()?;

        let chunk_type = extractor.capture_name_to_chunk_type(capture_name)?;

        let name = extractor
            .extract_name(node, source_code, capture_name)
            .or_else(|| Self::extract_name(node, source_code))
            .unwrap_or_else(|| "unknown".to_string());

        let chunk_comment_ranges: Vec<(usize, usize)> = file_comment_ranges
            .iter()
            .filter_map(|&(comment_start, comment_end)| {
                if comment_start >= start_byte && comment_end <= end_byte {
                    Some((comment_start - start_byte, comment_end - start_byte))
                } else {
                    None
                }
            })
            .collect();

        let normalized_content = Self::normalize_first_line_indentation(
            content,
            &original_indent_chars,
        );

        Some(CodeChunk {
            content: normalized_content,
            file_path: file_path.to_path_buf(),
            start_line,
            end_line,
            language: language.to_string(),
            chunk_type,
            name,
            comment_ranges: chunk_comment_ranges,
            original_indentation,
        })
    }

    fn extract_name(node: Node, source_code: &str) -> Option<String> {
        if node.kind() == "variable_declarator" {
            let mut cursor = node.walk();
            if cursor.goto_first_child() {
                let name_node = cursor.node();
                if name_node.kind() == "identifier" {
                    let start = name_node.start_byte();
                    let end = name_node.end_byte();
                    return Some(source_code[start..end].to_string());
                }
            }
        }

        let mut cursor = node.walk();
        if cursor.goto_first_child() {
            loop {
                let child = cursor.node();
                if child.kind() == "identifier"
                    || child.kind() == "type_identifier"
                    || child.kind() == "property_identifier"
                    || child.kind() == "field_identifier"
                    || child.kind() == "constant"
                    || child.kind() == "simple_identifier"
                {
                    let start = child.start_byte();
                    let end = child.end_byte();
                    return Some(source_code[start..end].to_string());
                }
                if !cursor.goto_next_sibling() {
                    break;
                }
            }
        }
        None
    }

    fn normalize_first_line_indentation(
        content: &str,
        original_indent_chars: &str,
    ) -> String {
        let lines: Vec<&str> = content.lines().collect();
        if lines.is_empty() {
            return content.to_string();
        }

        let mut result_lines = Vec::new();
        
        for (line_idx, line) in lines.iter().enumerate() {
            if line_idx == 0 {
                // First line: add original indentation characters from source
                result_lines.push(format!("{}{}", original_indent_chars, line));
            } else {
                // Other lines: keep as is
                result_lines.push(line.to_string());
            }
        }

        result_lines.join("\n")
    }

    fn extract_line_indent_chars(
        source_code: &str,
        line_row: usize,
        indent_length: usize,
    ) -> String {
        let lines: Vec<&str> = source_code.lines().collect();
        if line_row < lines.len() {
            let line = lines[line_row];
            line.chars().take(indent_length).collect()
        } else {
            String::new()
        }
    }
}
