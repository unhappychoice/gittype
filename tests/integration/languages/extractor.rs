// Don't import here since they're used inside the macro
// use these inside the macro expansion

/// Test macro for language extractors with snapshot functionality
macro_rules! test_language_extractor {
    (
        name: $test_name:ident,
        language: $language:literal,
        extension: $extension:literal,
        source: $source:expr,
        total_chunks: $total:expr,
        chunk_counts: {
            $($chunk_type:ident: $count:expr),* $(,)?
        }
    ) => {
        #[test]
        fn $test_name() {
            use crate::integration::{extract_chunks_for_test, test_extraction_options};
            use gittype::extractor::CodeChunkExtractor;
            use gittype::models::ChunkType;
            use std::collections::HashMap;
            use std::fs;
            use tempfile::TempDir;

            let temp_dir = TempDir::new().unwrap();
            let file_path = temp_dir.path().join(concat!("test.", $extension));

            fs::write(&file_path, $source).unwrap();

            let mut extractor = CodeChunkExtractor::new().unwrap();
            let chunks = extract_chunks_for_test(
                &mut extractor,
                temp_dir.path(),
                test_extraction_options()
            ).unwrap();

            // Check total chunk count
            assert_eq!(
                chunks.len(),
                $total,
                "Expected {} total chunks, got {}",
                $total,
                chunks.len()
            );

            // Check chunk counts by type
            let mut chunk_counts: HashMap<ChunkType, usize> = HashMap::new();
            for chunk in &chunks {
                *chunk_counts.entry(chunk.chunk_type.clone()).or_insert(0) += 1;
            }

            $(
                let expected_count = $count;
                let actual_count = chunk_counts.get(&ChunkType::$chunk_type).copied().unwrap_or(0);
                assert_eq!(
                    actual_count,
                    expected_count,
                    "Expected {} {:?} chunks, got {}",
                    expected_count,
                    ChunkType::$chunk_type,
                    actual_count
                );
            )*

            // Verify all chunks have correct language
            for chunk in &chunks {
                assert_eq!(chunk.language, $language.to_string());
            }

            // Always create snapshot
            crate::integration::languages::extractor::create_insta_snapshot(&chunks, $source, stringify!($test_name));
        }
    };
}

/// Create a snapshot using insta
pub fn create_insta_snapshot(
    chunks: &[gittype::models::CodeChunk],
    source_code: &str,
    test_name: &str,
) {
    use insta::assert_snapshot;
    use serde_json::json;

    let snapshot_data = json!({
        "test_name": test_name,
        "source_code": source_code,
        "total_chunks": chunks.len(),
        "chunks": chunks.iter().map(|chunk| {
            json!({
                "chunk_type": format!("{:?}", chunk.chunk_type),
                "name": chunk.name,
                "language": chunk.language,
                "start_line": chunk.start_line,
                "end_line": chunk.end_line,
                "original_indentation": chunk.original_indentation,
                "comment_ranges": chunk.comment_ranges,
                "content": chunk.content
            })
        }).collect::<Vec<_>>()
    });

    assert_snapshot!(
        test_name,
        serde_json::to_string_pretty(&snapshot_data).unwrap()
    );
}

// Export the macro for use in other modules
pub(crate) use test_language_extractor;
