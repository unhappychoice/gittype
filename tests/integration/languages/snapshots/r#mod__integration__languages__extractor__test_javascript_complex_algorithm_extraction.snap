---
source: tests/integration/languages/extractor.rs
expression: "serde_json::to_string_pretty(&snapshot_data).unwrap()"
---
{
  "chunks": [
    {
      "chunk_type": "Function",
      "comment_ranges": [
        [
          173,
          228
        ],
        [
          612,
          643
        ]
      ],
      "content": "function complexDataProcessor(input, options = {}) {\n    const { threshold = 10, enableCaching = true } = options;\n    const cache = new Map();\n    const results = [];\n\n    // Main processing algorithm - extractable middle chunk\n    for (let i = 0; i < input.length; i++) {\n        const item = input[i];\n        const cacheKey = `item_${i}_${item.id || i}`;\n\n        if (enableCaching && cache.has(cacheKey)) {\n            const cachedResult = cache.get(cacheKey);\n            if (cachedResult.valid) {\n                results.push(cachedResult.data);\n                continue;\n            }\n        }\n\n        // Complex transformation logic\n        let processedItem;\n        if (typeof item === 'object' && item !== null) {\n            const score = (item.value || 0) * (item.weight || 1);\n\n            if (score > threshold) {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score,\n                    category: score > threshold * 2 ? 'high' : 'medium',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'complex'\n                    }\n                };\n            } else {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score + threshold,\n                    category: 'low',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'simple',\n                        adjusted: true\n                    }\n                };\n            }\n        } else {\n            processedItem = {\n                id: i,\n                originalValue: item,\n                score: Number(item) || 0,\n                category: 'primitive',\n                metadata: {\n                    processed: true,\n                    timestamp: Date.now(),\n                    processor: 'primitive'\n                }\n            };\n        }\n\n        if (enableCaching) {\n            cache.set(cacheKey, { data: processedItem, valid: true });\n        }\n\n        results.push(processedItem);\n    }\n\n    return results;\n}",
      "end_line": 73,
      "language": "javascript",
      "name": "complexDataProcessor",
      "original_indentation": 0,
      "start_line": 2
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "        if (enableCaching && cache.has(cacheKey)) {\n            const cachedResult = cache.get(cacheKey);\n            if (cachedResult.valid) {\n                results.push(cachedResult.data);\n                continue;\n            }\n        }",
      "end_line": 18,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 8,
      "start_line": 12
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "            if (cachedResult.valid) {\n                results.push(cachedResult.data);\n                continue;\n            }",
      "end_line": 17,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 12,
      "start_line": 14
    },
    {
      "chunk_type": "CodeBlock",
      "comment_ranges": [],
      "content": "        if (typeof item === 'object' && item !== null) {\n            const score = (item.value || 0) * (item.weight || 1);\n\n            if (score > threshold) {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score,\n                    category: score > threshold * 2 ? 'high' : 'medium',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'complex'\n                    }\n                };\n            } else {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score + threshold,\n                    category: 'low',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'simple',\n                        adjusted: true\n                    }\n                };\n            }\n        }",
      "end_line": 51,
      "language": "javascript",
      "name": "code_block",
      "original_indentation": 55,
      "start_line": 22
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "        if (typeof item === 'object' && item !== null) {\n            const score = (item.value || 0) * (item.weight || 1);\n\n            if (score > threshold) {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score,\n                    category: score > threshold * 2 ? 'high' : 'medium',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'complex'\n                    }\n                };\n            } else {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score + threshold,\n                    category: 'low',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'simple',\n                        adjusted: true\n                    }\n                };\n            }\n        } else {\n            processedItem = {\n                id: i,\n                originalValue: item,\n                score: Number(item) || 0,\n                category: 'primitive',\n                metadata: {\n                    processed: true,\n                    timestamp: Date.now(),\n                    processor: 'primitive'\n                }\n            };\n        }",
      "end_line": 63,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 8,
      "start_line": 22
    },
    {
      "chunk_type": "CodeBlock",
      "comment_ranges": [],
      "content": "            if (score > threshold) {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score,\n                    category: score > threshold * 2 ? 'high' : 'medium',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'complex'\n                    }\n                };\n            }",
      "end_line": 37,
      "language": "javascript",
      "name": "code_block",
      "original_indentation": 35,
      "start_line": 25
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "            if (score > threshold) {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score,\n                    category: score > threshold * 2 ? 'high' : 'medium',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'complex'\n                    }\n                };\n            } else {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score + threshold,\n                    category: 'low',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'simple',\n                        adjusted: true\n                    }\n                };\n            }",
      "end_line": 50,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 12,
      "start_line": 25
    },
    {
      "chunk_type": "CodeBlock",
      "comment_ranges": [],
      "content": "            } else {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score + threshold,\n                    category: 'low',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'simple',\n                        adjusted: true\n                    }\n                };\n            }",
      "end_line": 50,
      "language": "javascript",
      "name": "code_block",
      "original_indentation": 19,
      "start_line": 37
    },
    {
      "chunk_type": "CodeBlock",
      "comment_ranges": [],
      "content": "        } else {\n            processedItem = {\n                id: i,\n                originalValue: item,\n                score: Number(item) || 0,\n                category: 'primitive',\n                metadata: {\n                    processed: true,\n                    timestamp: Date.now(),\n                    processor: 'primitive'\n                }\n            };\n        }",
      "end_line": 63,
      "language": "javascript",
      "name": "code_block",
      "original_indentation": 15,
      "start_line": 51
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "        if (enableCaching) {\n            cache.set(cacheKey, { data: processedItem, valid: true });\n        }",
      "end_line": 67,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 8,
      "start_line": 65
    },
    {
      "chunk_type": "Method",
      "comment_ranges": [
        [
          196,
          252
        ]
      ],
      "content": "    analyzePatterns(data, patternTypes = ['sequence', 'frequency']) {\n        const analysis = {\n            patterns: [],\n            statistics: {},\n            insights: []\n        };\n\n        // Pattern analysis algorithm - extractable middle chunk\n        if (patternTypes.includes('sequence')) {\n            for (let i = 1; i < data.length; i++) {\n                const current = data[i];\n                const previous = data[i - 1];\n\n                if (typeof current === 'number' && typeof previous === 'number') {\n                    const difference = current - previous;\n                    const percentChange = previous !== 0 ? (difference / previous) * 100 : 0;\n\n                    const pattern = {\n                        type: 'sequence',\n                        position: i,\n                        change: difference,\n                        percentChange: percentChange,\n                        trend: difference > 0 ? 'increasing' :\n                               difference < 0 ? 'decreasing' : 'stable',\n                        magnitude: Math.abs(percentChange) > 50 ? 'significant' : 'minor'\n                    };\n\n                    analysis.patterns.push(pattern);\n                }\n            }\n        }\n\n        if (patternTypes.includes('frequency')) {\n            const frequency = {};\n            data.forEach(item => {\n                const key = typeof item === 'object' ?\n                    JSON.stringify(item) : String(item);\n                frequency[key] = (frequency[key] || 0) + 1;\n            });\n\n            analysis.statistics.frequency = frequency;\n            analysis.statistics.mostCommon = Object.entries(frequency)\n                .sort(([,a], [,b]) => b - a)\n                .slice(0, 5);\n        }\n\n        return analysis;\n    }",
      "end_line": 123,
      "language": "javascript",
      "name": "analyzePatterns",
      "original_indentation": 4,
      "start_line": 76
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "        if (patternTypes.includes('sequence')) {\n            for (let i = 1; i < data.length; i++) {\n                const current = data[i];\n                const previous = data[i - 1];\n\n                if (typeof current === 'number' && typeof previous === 'number') {\n                    const difference = current - previous;\n                    const percentChange = previous !== 0 ? (difference / previous) * 100 : 0;\n\n                    const pattern = {\n                        type: 'sequence',\n                        position: i,\n                        change: difference,\n                        percentChange: percentChange,\n                        trend: difference > 0 ? 'increasing' :\n                               difference < 0 ? 'decreasing' : 'stable',\n                        magnitude: Math.abs(percentChange) > 50 ? 'significant' : 'minor'\n                    };\n\n                    analysis.patterns.push(pattern);\n                }\n            }\n        }",
      "end_line": 106,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 8,
      "start_line": 84
    },
    {
      "chunk_type": "Loop",
      "comment_ranges": [],
      "content": "            for (let i = 1; i < data.length; i++) {\n                const current = data[i];\n                const previous = data[i - 1];\n\n                if (typeof current === 'number' && typeof previous === 'number') {\n                    const difference = current - previous;\n                    const percentChange = previous !== 0 ? (difference / previous) * 100 : 0;\n\n                    const pattern = {\n                        type: 'sequence',\n                        position: i,\n                        change: difference,\n                        percentChange: percentChange,\n                        trend: difference > 0 ? 'increasing' :\n                               difference < 0 ? 'decreasing' : 'stable',\n                        magnitude: Math.abs(percentChange) > 50 ? 'significant' : 'minor'\n                    };\n\n                    analysis.patterns.push(pattern);\n                }\n            }",
      "end_line": 105,
      "language": "javascript",
      "name": "for_loop",
      "original_indentation": 12,
      "start_line": 85
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "                if (typeof current === 'number' && typeof previous === 'number') {\n                    const difference = current - previous;\n                    const percentChange = previous !== 0 ? (difference / previous) * 100 : 0;\n\n                    const pattern = {\n                        type: 'sequence',\n                        position: i,\n                        change: difference,\n                        percentChange: percentChange,\n                        trend: difference > 0 ? 'increasing' :\n                               difference < 0 ? 'decreasing' : 'stable',\n                        magnitude: Math.abs(percentChange) > 50 ? 'significant' : 'minor'\n                    };\n\n                    analysis.patterns.push(pattern);\n                }",
      "end_line": 104,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 16,
      "start_line": 89
    },
    {
      "chunk_type": "Conditional",
      "comment_ranges": [],
      "content": "        if (patternTypes.includes('frequency')) {\n            const frequency = {};\n            data.forEach(item => {\n                const key = typeof item === 'object' ?\n                    JSON.stringify(item) : String(item);\n                frequency[key] = (frequency[key] || 0) + 1;\n            });\n\n            analysis.statistics.frequency = frequency;\n            analysis.statistics.mostCommon = Object.entries(frequency)\n                .sort(([,a], [,b]) => b - a)\n                .slice(0, 5);\n        }",
      "end_line": 120,
      "language": "javascript",
      "name": "if_block",
      "original_indentation": 8,
      "start_line": 108
    },
    {
      "chunk_type": "Lambda",
      "comment_ranges": [],
      "content": "            data.forEach(item => {\n                const key = typeof item === 'object' ?\n                    JSON.stringify(item) : String(item);\n                frequency[key] = (frequency[key] || 0) + 1;\n            }",
      "end_line": 114,
      "language": "javascript",
      "name": "arrow_lambda",
      "original_indentation": 25,
      "start_line": 110
    },
    {
      "chunk_type": "FunctionCall",
      "comment_ranges": [],
      "content": "            analysis.statistics.mostCommon = Object.entries(frequency)\n                .sort(([,a], [,b]) => b - a)",
      "end_line": 118,
      "language": "javascript",
      "name": "function_call",
      "original_indentation": 45,
      "start_line": 117
    },
    {
      "chunk_type": "FunctionCall",
      "comment_ranges": [],
      "content": "            analysis.statistics.mostCommon = Object.entries(frequency)\n                .sort(([,a], [,b]) => b - a)\n                .slice(0, 5)",
      "end_line": 119,
      "language": "javascript",
      "name": "function_call",
      "original_indentation": 45,
      "start_line": 117
    }
  ],
  "source_code": "\nfunction complexDataProcessor(input, options = {}) {\n    const { threshold = 10, enableCaching = true } = options;\n    const cache = new Map();\n    const results = [];\n\n    // Main processing algorithm - extractable middle chunk\n    for (let i = 0; i < input.length; i++) {\n        const item = input[i];\n        const cacheKey = `item_${i}_${item.id || i}`;\n\n        if (enableCaching && cache.has(cacheKey)) {\n            const cachedResult = cache.get(cacheKey);\n            if (cachedResult.valid) {\n                results.push(cachedResult.data);\n                continue;\n            }\n        }\n\n        // Complex transformation logic\n        let processedItem;\n        if (typeof item === 'object' && item !== null) {\n            const score = (item.value || 0) * (item.weight || 1);\n\n            if (score > threshold) {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score,\n                    category: score > threshold * 2 ? 'high' : 'medium',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'complex'\n                    }\n                };\n            } else {\n                processedItem = {\n                    id: item.id || i,\n                    originalValue: item.value,\n                    score: score + threshold,\n                    category: 'low',\n                    metadata: {\n                        processed: true,\n                        timestamp: Date.now(),\n                        processor: 'simple',\n                        adjusted: true\n                    }\n                };\n            }\n        } else {\n            processedItem = {\n                id: i,\n                originalValue: item,\n                score: Number(item) || 0,\n                category: 'primitive',\n                metadata: {\n                    processed: true,\n                    timestamp: Date.now(),\n                    processor: 'primitive'\n                }\n            };\n        }\n\n        if (enableCaching) {\n            cache.set(cacheKey, { data: processedItem, valid: true });\n        }\n\n        results.push(processedItem);\n    }\n\n    return results;\n}\n\nconst dataAnalyzer = {\n    analyzePatterns(data, patternTypes = ['sequence', 'frequency']) {\n        const analysis = {\n            patterns: [],\n            statistics: {},\n            insights: []\n        };\n\n        // Pattern analysis algorithm - extractable middle chunk\n        if (patternTypes.includes('sequence')) {\n            for (let i = 1; i < data.length; i++) {\n                const current = data[i];\n                const previous = data[i - 1];\n\n                if (typeof current === 'number' && typeof previous === 'number') {\n                    const difference = current - previous;\n                    const percentChange = previous !== 0 ? (difference / previous) * 100 : 0;\n\n                    const pattern = {\n                        type: 'sequence',\n                        position: i,\n                        change: difference,\n                        percentChange: percentChange,\n                        trend: difference > 0 ? 'increasing' :\n                               difference < 0 ? 'decreasing' : 'stable',\n                        magnitude: Math.abs(percentChange) > 50 ? 'significant' : 'minor'\n                    };\n\n                    analysis.patterns.push(pattern);\n                }\n            }\n        }\n\n        if (patternTypes.includes('frequency')) {\n            const frequency = {};\n            data.forEach(item => {\n                const key = typeof item === 'object' ?\n                    JSON.stringify(item) : String(item);\n                frequency[key] = (frequency[key] || 0) + 1;\n            });\n\n            analysis.statistics.frequency = frequency;\n            analysis.statistics.mostCommon = Object.entries(frequency)\n                .sort(([,a], [,b]) => b - a)\n                .slice(0, 5);\n        }\n\n        return analysis;\n    }\n};\n",
  "test_name": "test_javascript_complex_algorithm_extraction",
  "total_chunks": 18
}
